src/protocols/secure_channel/PASESession.cpp

```diff
@@ -700,56 +789,65 @@ CHIP_ERROR PASESession::HandleMsg3(const System::PacketBufferHandle & msg)
+    StatusReport statusReport(generalCode, protocolId, protocolCode);
 
-    pMsg        = reinterpret_cast<Spake2pErrorMsg *>(msg->Start());
-    pMsg->error = errorCode;
+    Encoding::LittleEndian::PacketBufferWriter bbuf(System::PacketBufferHandle::New(statusReport.Size()));
+    statusReport.WriteToBuffer(bbuf);
 
-    msg->SetDataLength(msglen);
+    System::PacketBufferHandle msg = bbuf.Finalize();
+    VerifyOrReturn(!msg.IsNull(), ChipLogError(SecureChannel, "Failed to allocate status report message"));
 
-    VerifyOrReturn(mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::PASE_Spake2pError, std::move(msg)) ==
-                       CHIP_NO_ERROR,
-                   ChipLogError(SecureChannel, "Failed to send error message"));
+    VerifyOrReturn(mExchangeCtxt->SendMessage(MsgType::StatusReport, std::move(msg)) == CHIP_NO_ERROR,
+                   ChipLogError(SecureChannel, "Failed to send status report message"));
 }
 
-CHIP_ERROR PASESession::HandleErrorMsg(const System::PacketBufferHandle & msg)
+CHIP_ERROR PASESession::HandleStatusReport(System::PacketBufferHandle && msg)
 {
-    ReturnErrorCodeIf(msg->Start() == nullptr || msg->DataLength() != sizeof(Spake2pErrorMsg), CHIP_ERROR_MESSAGE_INCOMPLETE);
-
-    static_assert(
-        sizeof(Spake2pErrorMsg) == sizeof(uint8_t),
-        "Assuming size of Spake2pErrorMsg message is 1 octet, so that endian-ness conversion and memory alignment is not needed");
+    StatusReport report;
+    CHIP_ERROR err = report.Parse(std::move(msg));
+    ReturnErrorOnFailure(err);
 
-    Spake2pErrorMsg * pMsg = reinterpret_cast<Spake2pErrorMsg *>(msg->Start());
-
-    CHIP_ERROR err = CHIP_NO_ERROR;
-    switch (pMsg->error)
+    if (report.GetGeneralCode() == GeneralStatusCode::kSuccess && report.GetProtocolCode() == kProtocolCodeSuccess)
     {
-    case Spake2pErrorType::kInvalidKeyConfirmation:
-        err = CHIP_ERROR_KEY_CONFIRMATION_FAILED;
-        break;
+        mPairingComplete = true;
 
-    case Spake2pErrorType::kUnexpected:
-        err = CHIP_ERROR_INVALID_PASE_PARAMETER;
-        break;
+        // Forget our exchange, as no additional messages are expected from the peer
+        mExchangeCtxt = nullptr;
 
-    default:
-        err = CHIP_ERROR_INTERNAL;
-        break;
-    };
-    ChipLogError(SecureChannel, "Received error during pairing process. %s", ErrorStr(err));
+        // Call delegate to indicate pairing completion
+        mDelegate->OnSessionEstablished();
+    }
+    else
+    {
+        switch (report.GetProtocolCode())
+        {
+        case kProtocolCodeNoSharedRoot:
+            err = CHIP_ERROR_KEY_CONFIRMATION_FAILED;
+            break;
+
+        case kProtocolCodeInvalidParam:
+            err = CHIP_ERROR_INVALID_PASE_PARAMETER;
+            break;
+
+        default:
+            err = CHIP_ERROR_INTERNAL;
+            break;
+        };
+        ChipLogError(SecureChannel, "Received error during pairing process. %s", ErrorStr(err));
```

Comment:
Should also log the original protocolCode since we are doing "flattening" error code conversion w/ a default fallback.